ocf_data_sampler/torch_datasets/site.py

```diff
@@ -0,0 +1,271 @@
+    # Should we have a different option if there are not nans
+    sites = datasets_dict["site"]
+    site_ids = sites.site_id.values
+    site_config = config.input_data.site
+    valid_t0_and_site_ids = []
+    for site_id in site_ids:
+        site = sites.sel(site_id=site_id)
+
+        # drop any nan values
+        # not sure this is right?
+        site = site.dropna(dim='time_utc')
+
+        # Get the valid time periods for this location
+        time_periods = find_contiguous_t0_periods(
+            pd.DatetimeIndex(site["time_utc"]),
+            sample_period_duration=minutes(site_config.time_resolution_minutes),
+            history_duration=minutes(site_config.history_minutes),
+            forecast_duration=minutes(site_config.forecast_minutes),
+        )
+        valid_time_periods_per_site = intersection_of_multiple_dataframes_of_periods(
+            [valid_time_periods, time_periods]
+        )
+
+        # Fill out the contiguous time periods to get the t0 times
+        valid_t0_times_per_site = fill_time_periods(
+            valid_time_periods_per_site,
+            freq=minutes(site_config.time_resolution_minutes)
+        )
+
+        valid_t0_per_site = pd.DataFrame(index=valid_t0_times_per_site)
+        valid_t0_per_site['site_id'] = site_id
+        valid_t0_and_site_ids.append(valid_t0_per_site)
+
+    valid_t0_and_site_ids = pd.concat(valid_t0_and_site_ids)
+    valid_t0_and_site_ids.index.name = 't0'
+    valid_t0_and_site_ids.reset_index(inplace=True)
+
+    return valid_t0_and_site_ids
+
+
+def get_locations(site_xr:xr.Dataset):
+    """Get list of locations of all sites"""
+
+    locations = []
+    for site_id in site_xr.site_id.values:
+        site = site_xr.sel(site_id=site_id)
+        location = Location(
+            id=site_id,
+            x=site.longitude.values,
+            y=site.latitude.values,
+            coordinate_system="lon_lat"
+        )
+        locations.append(location)
+
+    return locations
+
+
+class SitesDataset(Dataset):
+    def __init__(
+            self,
+            config_filename: str,
+            start_time: str | None = None,
+            end_time: str | None = None,
+            gsp_ids: list[int] | None = None,
+    ):
+        """A torch Dataset for creating PVNet UK GSP samples
+
+        Args:
+            config_filename: Path to the configuration file
+            start_time: Limit the init-times to be after this
+            end_time: Limit the init-times to be before this
+            gsp_ids: List of GSP IDs to create samples for. Defaults to all
+        """
+
+        config = load_yaml_configuration(config_filename)
+
+        datasets_dict = get_dataset_dict(config)
+
+        # get all locations
+        self.locations = get_locations(datasets_dict['site'])
+
+        # Get t0 times where all input data is available
+        valid_t0_and_site_ids = find_valid_t0_and_site_ids(datasets_dict, config)
+
+        # Filter t0 times to given range
```

Comment:
Think this comment can be removed

Reply:
Ive added in those ranges now, so should be good to use + added some tests